File: src/main/java/recursive/PascalTriangle.java
package recursive;

import java.util.ArrayList;
import java.util.List;

/**
 * 给定一个非负整数 numRows，生成杨辉三角的前 numRows 行。
 *
 * 示例：
 * 输入: 5
 * 输出:
 * [
 *      [1],
 *     [1,1],
 *    [1,2,1],
 *   [1,3,3,1],
 *  [1,4,6,4,1]
 * ]
 */
public class PascalTriangle {
    public static void main(String args[]){
        List<List<Integer>> triangle = generate(5);
        for (List<Integer> line : triangle) {
            System.out.println(line);
        }
    }

    public static List<List<Integer>> generate(int numRows) {
        List<List<Integer>> res = new ArrayList<List<Integer>>(numRows);
        for (int i = 0; i < numRows; i++) {
            List<Integer> list = new ArrayList<Integer>(i + 1);
            for (int j = 0; j <= i; j++) {
                Integer val = calc(i, j);
                list.add(val);
            }
            res.add(list);
        }
        return res;
    }

    /**
     * 递归计算帕斯卡三角的第(i, j)个数值，如同计算斐波拉契数列第n个值一样，这里会涉及到重复计算
     */
    private static Integer calc(int i, int j) {
        if (j == 0 || i == j)
            return 1;
        return calc(i-1, j-1) + calc(i-1, j);
    }
}

File: CppAlgorithms/recursive/PascalTriangle.cpp
#include <vector>
#include <iostream>
using namespace std;

/**
 * https://leetcode-cn.com/problems/pascals-triangle/
 *
 * 给定一个非负整数 numRows，生成杨辉三角的前 numRows 行。
 *
 * 示例：
 * 输入: 5
 * 输出:
 * [
 *      [1],
 *     [1,1],
 *    [1,2,1],
 *   [1,3,3,1],
 *  [1,4,6,4,1]
 * ]
 */
class PascalTriangle {
private:
    vector<vector<int>> *res;

    void recursive(int numRows, vector<int> *before){
        --numRows;
        if (numRows == 0){
            return;
        }
        vector<int> *later = new vector<int>;
        for (int i = 0; i < before->size()+1; ++i) {
            int left;
            int right;
            if (i == 0)
                left = 0;
            else
                left = (*before)[i-1];
            if (i == before->size())
                right = 0;
            else
                right = (*before)[i];
            later->push_back(left + right);
        }
        res->push_back(*later);
        recursive(numRows, later);
    }

public:
    vector<vector<int>> generate(int numRows) {
        res = new vector<vector<int>>;
        if (numRows == 0){
            return *res;
        }
        auto *first = new vector<int>{1};
        res->push_back(*first);
        recursive(numRows, first);
        return *res;
    }
};

int main(){
    PascalTriangle s{};
    vector<vector<int>> res = s.generate(6);
    for(vector<int> v: res){
        for (int e: v){
            cout << e << " ";
        }
        cout << endl;
    }
}